File: js/howSum_tab.js
// howSum(targetSum, numbers)
//
// howSum(7, [5,3,4]) -> [4,3]

function howSum(targetSum, numbers) {
	const table = Array(targetSum + 1).fill(null)
	table[0] = []

	for (let i=0; i <= targetSum; i++) {
		if (table[i] !== null)
			for (let num of numbers) {
				if (i + num <= targetSum)
					table[i + num] = [...table[i], num]
			}
	}

	return table[targetSum]
}

console.log(howSum(7, [3, 4, 5]))
console.log(howSum(8, [3, 4, 5]))
console.log(howSum(7, [2, 4]))
console.log(howSum(13, [3, 4, 5]))
console.log(howSum(300, [7, 14]))

File: cpp/howSum_tab.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

vector<int> howSum(int target, const vector<int>& numbers) {
	vector<vector<int>> table (target + 1, vector<int>());
	table[0] = {0};

	for (int i=0; i<=target; i++) {
		if (!table[i].empty()) {
			if (table[i][0] == 0) table[i].pop_back();
			for (auto num : numbers) {
				if (i + num <= target) {
					table[num + i] = table[i];
					table[num + i].push_back(num);
				}
			}
		}
	}

	return table[target];
}

vector<int> howSum(int target, const vector<int>&& numbers) {
	return howSum(target, numbers);
}

int main() {
	auto result = howSum(7, {3, 4, 7});
	for (auto r : result) {
		cout << r << " ";
	}
	cout << endl;
}
